src/api/app/controllers/webui/users/tokens_controller.rb

```diff
@@ -105,19 +106,28 @@ def set_package
     # Check if only project_name or only package_name are present
     if @extra_params[:project_name].present? ^ @extra_params[:package_name].present?
       flash[:error] = 'When providing an optional package, both Project name and Package name must be provided.'
-      redirect_to new_token_path and return
+      @token = reinitialize_token(@params)
+      render :new and return
     end
 
     # If both project_name and package_name are present, check if this is an existing package
     begin
       @package = Package.get_by_project_and_name(@extra_params[:project_name], @extra_params[:package_name])
     rescue Project::UnknownObjectError
       flash[:error] = "When providing an optional package, the package must exist. Project '#{elide(@extra_params[:project_name])}' does not exist."
-      redirect_to new_token_path
+      @token = reinitialize_token(@params)
+      render :new
     rescue Package::UnknownObjectError
       flash[:error] = 'When providing an optional package, the package must exist. ' \
                       "Package '#{elide(@extra_params[:project_name])}/#{elide(@extra_params[:package_name])}' does not exist."
-      redirect_to new_token_path
+      @token = reinitialize_token(@params)
+      render :new
     end
   end
+
+  def reinitialize_token(params)
```

Comment:
I know what `Token.new` does, I have to look up what `reinitialize_token` does. I don't think it's worth turning this into a method...